langgraph/src/checkpoint/sqlite.ts

```diff
@@ -0,0 +1,195 @@
+import Database, { Database as DatabaseType } from "better-sqlite3";
+import { RunnableConfig } from "@langchain/core/runnables";
+import {
+  BaseCheckpointSaver,
+  Checkpoint,
+  CheckpointAt,
+  CheckpointTuple,
+  SerializerProtocol,
+} from "./base.js";
+
+export class JsonSerializer implements SerializerProtocol {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  dumps(obj: any): string {
+    return JSON.stringify(obj);
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  loads(data: string): any {
+    return JSON.parse(data);
+  }
+}
+
+// snake_case is used to match Python implementation
+interface Row {
+  checkpoint: string;
+  parent_ts: string;
+  thread_id?: string;
+  thread_ts?: string;
+}
+
+export class SqliteSaver extends BaseCheckpointSaver {
+  serde = new JsonSerializer();
+
+  db: DatabaseType;
+
+  constructor(
+    connStringOrLocalPath: string,
+    serde?: SerializerProtocol,
+    at?: CheckpointAt
+  ) {
+    super(serde, at);
+    this.db = new Database(connStringOrLocalPath);
+    this.setup().catch((error) =>
```

Comment:
isn't there a race condition here? you could end up calling one of the methods before setup finishes

Reply:
Yes, you're right. Let me fix.

I reproduced the race condition by adding a sleep statement.
```typescript
private async setup(): Promise<void> {
  // sleep for 3 seconds
  await new Promise((resolve) => setTimeout(resolve, 3000));
  ...
}
```
The unit tests failed.